File: Engine/src/Modules/Stellar.Core/Label/Label.cs
using Stellar.Kernel;
using Stellar.Kernel.Quantization;
using Stellar.Kernel.Label;
using Stellar.Kernel.Data.Registry;
using Stellar.Core.Quantization;
using Stellar.Core.Data.Registry;

namespace Stellar.Core.Label;

/// <inheritdoc cref="ILabel" />
public sealed class Label
    : MetaQuant, ILabel
{
    /// <inheritdoc/>
    public string Name { get; }

    private Label(IIdentifier identifier, string name)
        : base(identifier)
    {
        Name = name;
    }

    /// <inheritdoc/>
    public void Register(IQuantumObject? registry = null)
    {
        registry ??= LabelRegistry.Instance;
        if (registry is IRegistry<ILabel> identifierRegistry)
            identifierRegistry.Register(this);
    }

    /// <inheritdoc/>
    public override string ToString() => $"Label#{Name}";

    /// <inheritdoc/>
    public void Unregister(IQuantumObject? registry = null)
    {
        registry ??= LabelRegistry.Instance;
        if (registry is IRegistry<ILabel> identifierRegistry)
            identifierRegistry.Pop(UID);
    }

    /// <inheritdoc/>
    public void Dispose() => Unregister();

    #region Static Methoods

    public static Label? Create(string name, IIdentifier? identifier = null, bool register = true)
    {
        if (string.IsNullOrWhiteSpace(name))
            return null;

        identifier ??= new Identifier(Guid.NewGuid());
        var label = new Label(identifier, name);
        if (register) LabelRegistry.Instance.Register(label);
        return label;
    }

    public static Label? Get(IIdentifier identifier) => LabelRegistry.Instance.Get(identifier);
    public static Label? Get(string name) => LabelRegistry.Instance.GetByName(name);

    #endregion

    public static implicit operator Label?(string name) => Get(name) ?? Create(name);
}

File: Engine/src/Native/Stellar.Native.Core/Label.cpp
#include "pch.h"
#include "Label.h"

namespace Stellar::Native::Core
{
    Label Label::CreateBound(std::string_view name)
    {
        return Label{name, Identifier::Create()};
    }

    Label Label::CreateBound(std::string_view name, const Identifier& id_)
    {
        return Label{name, id_};
    }

    bool Label::IsValid() const noexcept
    {
        return !_name.empty();
    }

    bool Label::IsBound() const noexcept
    {
        return !_uid.IsNull();
    }

    bool Label::IsBound(const Identifier& id) const noexcept
    {
        return _uid == id;
    }
}
